vm/src/object/ext.rs

```diff
@@ -323,6 +317,46 @@ where
     }
 }
 
+pub struct PyAtomicRef<T: PyObjectPayload> {
+    inner: AtomicCell<PyRef<T>>,
+    // ptr could possible point to old or new inner, but guarentee the ref is valid
+    ptr: AtomicCell<NonNull<Py<T>>>,
+}
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "threading")] {
+        unsafe impl<T: Send + PyObjectPayload> Send for PyAtomicRef<T> {}
+        unsafe impl<T: Sync + PyObjectPayload> Sync for PyAtomicRef<T> {}
+    }
+}
+
+impl<T: PyObjectPayload> From<PyRef<T>> for PyAtomicRef<T> {
+    fn from(inner: PyRef<T>) -> Self {
+        let ptr = NonNull::from(inner.deref());
+        Self {
+            inner: AtomicCell::new(inner),
+            ptr: AtomicCell::new(ptr),
+        }
+    }
+}
+
+impl<T: PyObjectPayload> Deref for PyAtomicRef<T> {
+    type Target = Py<T>;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { self.ptr.load().as_ref() }
```

Comment:
This is looking unsafe.

While thread 1 called `deref()` to obtain `&Py<T>`, `inner` can be swapped out in thread 2.
When the swapped out object is dropped, thread 1 will have a dangling pointer.